File: cpp-lang/design-patterns/hf/CompositePatterns/ducksimulator/flock.cpp
#include "flock.h"

Flock::~Flock()
{
    for (Quackable *q : _quackers)
        delete q;
    _quackers.clear();
}

void Flock::add(Quackable *quacker)
{
    _quackers.push_back(quacker);
}

void Flock::quack() const
{
    for (Quackable *q : _quackers)
        q->quack();
}

void Flock::registerObserver(Observer *observer)
{
    for (Quackable *q : _quackers)
        q->registerObserver(observer);
}

void Flock::notifyObservers() const
{
    // Nothing to do
}

std::string Flock::name() const
{
    return std::string();
}

File: java-lang/design-patterns/CompositePatterns/DuckSimulator/src/Flock.java
import java.util.ArrayList;
import java.util.Iterator;

/**
 * Created by vova on 11/27/16.
 */
public class Flock implements Quackable {
    ArrayList quackers = new ArrayList();

    public void add(Quackable quacker) {
        quackers.add(quacker);
    }

    public void quack() {
        Iterator iterator = quackers.iterator();
        while (iterator.hasNext()) {
            Quackable quacker = (Quackable)iterator.next();
            quacker.quack();
        }
    }

    public void registerObserver(Observer observer) {
        Iterator iterator = quackers.iterator();
        while (iterator.hasNext()) {
            Quackable quacker = (Quackable)iterator.next();
            quacker.registerObserver(observer);
        }
    }

    public void notifyObservers() {
        // Nothing to do.
    }
}
